pkg/controllers/nodeclass/validation.go

```diff
@@ -324,19 +339,98 @@ func mockLaunchTemplateConfig() []ec2types.FleetLaunchTemplateConfigRequest {
+			continue
+		}
+		amiRequirements := scheduling.NewNodeSelectorRequirements(ami.Requirements...)
+		if amiRequirements.IsCompatible(amifamily.VariantStandard.Requirements()) {
+			instanceTypes = append(instanceTypes, amiMap[ami.ID]...)
+		}
+	}
+	if len(instanceTypes) == 0 && len(amiMap) != 0 {
+		instanceTypes = lo.Flatten(lo.Values(amiMap))
+	}
+	// If there weren't any matching instance types, we should fallback to some defaults. There's an instance type included
+	// for both x86_64 and arm64 architectures, ensuring that there will be a matching AMI.
+	if len(instanceTypes) == 0 {
+		instanceTypes = []*cloudprovider.InstanceType{
+			{
+				Name: string(ec2types.InstanceTypeM5Large),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureAmd64),
+				)...),
+			},
+			{
+				Name: string(ec2types.InstanceTypeM6gLarge),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureArm64),
+				)...),
+			},
+		}
+	}
+	opts, err := v.amiResolver.Resolve(nodeClass, nodeClaim, instanceTypes, karpv1.CapacityTypeOnDemand, amiOptions)
+	if err != nil {
+		return nil, err
+	}
+	return opts[0], nil
+}
+
+// getInstanceTypesForNodeClass returns the set of instances which could be launched using this NodeClass based on the
+// requirements of linked NodePools. If no NodePools exist for the given NodeClass, this function returns two default
+// instance types (one x86_64 and one arm64).
+func (v *Validation) getInstanceTypesForNodeClass(ctx context.Context, nodeClass *v1.EC2NodeClass) []*cloudprovider.InstanceType {
+	instanceTypes, err := v.cloudProvider.GetInstanceTypes(ctx, &karpv1.NodePool{Spec: karpv1.NodePoolSpec{Template: karpv1.NodeClaimTemplate{Spec: karpv1.NodeClaimTemplateSpec{
+		NodeClassRef: &karpv1.NodeClassReference{
+			Group: object.GVK(nodeClass).Group,
+			Kind:  object.GVK(nodeClass).Kind,
+			Name:  nodeClass.Name,
 		},
-	}, karpv1.CapacityTypeOnDemand, amiOptions)
+	}}}})
+	if err != nil {
+		return nil
+	}
+
+	nodePools, err := nodepool.ListManaged(ctx, v.kubeClient, v.cloudProvider, nodepool.ForNodeClass(nodeClass))
+	if err != nil {
+		return nil
+	}
+	var compatibleInstanceTypes []*cloudprovider.InstanceType
+	names := sets.New[string]()
+	for _, np := range nodePools {
+		reqs := scheduling.NewNodeSelectorRequirementsWithMinValues(np.Spec.Template.Spec.Requirements...)
+		for _, it := range instanceTypes {
+			if !it.Requirements.IsCompatible(reqs, scheduling.AllowUndefinedWellKnownLabels) {
+				continue
+			}
+			if names.Has(it.Name) {
+				continue
+			}
+			names.Insert(it.Name)
+			compatibleInstanceTypes = append(compatibleInstanceTypes, it)
+		}
+	}
+	if len(compatibleInstanceTypes) == 0 {
```

Comment:
nit: Do you really need to explicitly return nil if you have an empty array?

Reply:
Probably not, at one point this function returned an error and I never removed the conditional. Left it as is because it's a little more explicit, but I'm happy to drop it.

Reply:
Yeah, I'd probably just drop it honestly